### DISCUSSION TITLE
Release timeline for RU and PU

### DISCUSSION THREAD

**Discussion by oliver-steinbrecher** [Digitaler-Impfnachweis/certification-apis#66] [General]
### RU Environment

- 31.05.2021: Deployment of API [Release 2.0.0](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/2.0.0)
- 09.06.2021: Deployment of API [Release 2.1.0](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/2.1.0)

### PU Environment
(Regular maintenance windows: daily 20:00-21:00 UTC+2)

- 14.06.2021: Deployment of API [Release 2.0.0](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/2.0.0) 
- 22.06.2021: Deployment of API [Release 2.1.0](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/2.1.0)

**Comment by mge-medisoftware** [comment]
Nur mal zur Info: Unsere Hotline platzt gerade, die Praxen rufen uns an, weil sie wiederum von Patienten angerufen werden, die Sache mit den Impfnachweisen würde ja ab heute starten (wie in den Medien/Pressemeldungen kommuniziert).
Wir haben sogar schon ein Update rausgehauen, das wohl funktionieren würde (jedoch Funktion noch mit Passwort geschützt).
Da ist so richtig Druck drauf....

**Reply by dirkexe** [reply]
Ob ich an dieser Stelle etwas bewirken kann, bezweifele ich. Aber ich muss es loswerden: 

Die Apotheken werden derzeit gefühlt zumindest medial bevorzugt. Viel Aufwand wird hier getrieben, die Apotheken in die Lage zu versetzen, über eine Webseite die Impfzertifikate auszustellen. Apotheken haben noch nie einen Patienten geimpft und sollen über fremde Leistungen Zeugnis ablegen. 

Die Impfzentren werden die Zertifikate für erfolgte Impfungen automatisch verschicken.

In den Arztpraxen liegen die Daten zu den in der jeweiligen Praxis durchgeführten Impfungen strukturiert vor. Die Arztpraxen wissen, wer wann womit geimpft wurde - fälschungssicher. Auch wenn die Arztpraxen nur ein Bruchteil der Vergütung einer Apotheke für ein Zertifikat erhalten, sind die Arztpraxen doch bereit, die Zertifikate rückwirkend und in Zukunft direkt bei Impfung auszustellen.

Daher meine freundliche Bitte: **PU API heute - wie angekündigt - freischalten**. Dann können die Mitarbeiter der PVS-Hersteller sich um die Optimierung der Prozesse statt um Vertrösten kümmern.

Dirk Fellenberg | InterData Praxiscomputer GmbH

**Reply by dirkexe** [reply]
Nachdem der erste Termin für das PU Environment, der 10.06.2021 gerissen wurde, da das System offensichtlich weiterhin nicht verfügbar ist, frage ich mich, ob am 17.06.2021 für Release 2.1.0 festgehalten wird. Und falls ja, ob damit 17.06.2021 00:00 Uhr gemeint ist?

**Comment by oliver-steinbrecher** [comment]
Produktion wurde am Wochenende eingerichtet und steht nun zur Verfügung.

**Reply by dirkexe** [reply]
Der von den Arztpraxen laut [Dokumentation](https://github.com/Digitaler-Impfnachweis/certification-apis/blob/master/SMCB-Authentication.md) zu verwendende Endpunkt `https://id.impfnachweis.info/auth/realms/bmg-ti-certify/protocol/openid-connect/auth` antwortet aktuell in der PU unerwartet mit einem 400er-Fehler und einer HTML-Seite (vgl. auch [#87](https://github.com/Digitaler-Impfnachweis/certification-apis/discussions/87#discussioncomment-866801)). Wo können wir den Fehler suchen (finden)?

**Reply by oliver-steinbrecher** [reply]
Wir sind in der Analyse .

**Reply by oliver-steinbrecher** [reply]
Wir haben eine Korrektur ausgeliefert. Bitte erneut prüfen

**Reply by dirkexe** [reply]
Schritt 1 funktioniert jetzt. Wir bekommen die Challenge. Neuer Fehler im Schritt: "Submit Signed challenge and SMCB public certificate" reagiert mit 401 Unauthorized. Wir haben das zwei Praxen in der PU prüfen lassen.

**Reply by mge-medisoftware** [reply]
Bei unseren Tests haben wir ein identisches Fehlerverhalten. Http-Status 401 beim Senden der signierten challenge.

**Reply by oliver-steinbrecher** [reply]
Welche SMCB OIDs kommen zum Einsatz ?

**Reply by dirkexe** [reply]
`1.2.276.0.76.4.50`

Ich habe das gerade geprüft beim Kunden. Das ist eine typische Medisign SMC-B einer Arztpraxis ("Betriebsstätte Arzt").

**Reply by mge-medisoftware** [reply]
Alle Praxen müssten die 1.2.276.0.76.4.50 haben, (nicht ärztliche) Psychotherapeuten die 1.2.276.0.76.4.52, gematik-Betriebsstätten für tests o.ä. die 1.2.276.0.76.4.58

**Reply by dirkexe** [reply]
Jetzt kommt entweder erneut Fehler 400 Bad Request beim Schritt 1, dem Versuch, die Challenge abzurufen (siehe [oben](https://github.com/Digitaler-Impfnachweis/certification-apis/discussions/66#discussioncomment-866983)) ODER Fehler 403 Forbidden beim letzten Schritt `/api/certify/v2/issue` mit `Accept: application/cbor+base45`. Es geht also weiter! Ich versuche es später noch mal. Wenn wir mit Details helfen können, gern. Und wenn jemand meine Beobachtung bestätigen kann, auch gern.

**Reply by oliver-steinbrecher** [reply]
Wir haben um 18:03 eine Korrektur im Zusammenspiel mit der TSL ausgeliefert. @dirkexe können Sie nochmal prüfen?

**Reply by dirkexe** [reply]
Fehler 403 beim /api/certify/v2/issue

#### Request (gekürzt/geschwärzt):
```
POST /api/certify/v2/issue HTTP/1.1
Host: api.impfnachweis.info
Content-Type: application/vnd.dgc.v1+json
Content-Length: 327
Authorization: Bearer eyJh [...] MRA
Accept: application/cbor+base45

{
  "nam" : {
    "fn" : "xxx",
    "gn" : "xxx"
  },
  "dob" : "1xxx-xx-xx",
  "v" : [
    {
      "id" : "182084900",
      "tg" : "840539006",
      "vp" : "1119349007",
      "mp" : "EU/1/20/1528",
      "ma" : "ORG-100030215",
      "dn" : 1,
      "sd" : 2,
      "dt" : "2021-06-08"
    }
  ]
}

```

#### Response (gekürzt):
```
HTTP/1.1 403 Forbidden
Content-Type: text/plain;charset=utf-8
Date: Mon, 14 Jun 2021 16:23:43 GMT
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Server: istio-envoy
X-Envoy-Decorator-Operation: certifier-api.ubirch-prod.svc.cluster.local:8081/*

Sorry, the credential is invalid:Identities Not Found
```

**Reply by ManuelB** [reply]
@dirkexe dann scheint es aber schonmal geklappt zu haben einen Token zu bekommen :-)

**Reply by oliver-steinbrecher** [reply]
Ich sehe auch erfolgreiche Logins

**Reply by ManuelB** [reply]
Ich konnte gerade mit unserer Innovationspraxis web.impfnachweis.info erfolgreich testen.

**Reply by dirkexe** [reply]
Ich habe meinen letzten Post noch ausgeschmückt (siehe edit). Könnte die `id` noch gegen eine Whitelist geprüft werden? Sowas hatten wir zuerst auch in der RU?

> Sorry, the credential is invalid:Identities Not Found

**Reply by oliver-steinbrecher** [reply]
Wir haben eine erneute Anpassung geliefert. Lauffähigkeit ist bereits bestätigt worden

**Reply by dirkexe** [reply]
Ich kann das bestätigen. Der Prozess klappt jetzt. Danke!

**Reply by mge-medisoftware** [reply]
Ja, Prozess funktioniert jetzt. Server ist scheinbar aber manchmal überlastet (wie ja gestern auch von Apotheken gemeldet)? 504: gateway timeout bei api.impfnachweis.info
Wie hoch sollte man denn eigentlich die receive timeouts setzten (egal ob asynchroner oder synchrone Abruf)?

**Reply by dirkexe** [reply]
Bis etwa heute 9 Uhr hat es funktioniert, danach nicht mehr. Ein Fehler 504 tritt bei `/api/certify/v2/issue` auf. Wir haben die Abfrage bisher synchron mit einem Timeout von 30 Sekunden implementiert.

Auch die manuelle Eingabe über Web ohne API `https://web.impfnachweis.info` ist vom Timeout betroffen und meldet "Zeitüberschreitung bei der Erfassung der Daten. Daten konnten nicht erfasst werden. Bitte versuchen Sie es erneut."

**Reply by ManuelB** [reply]
Ich habe um 9:54 die Nachricht bekommen, dass https://web.impfnachweis.info nicht mehr funktioniert.

**Reply by dirkexe** [reply]
Update: Inzwischen funktioniert die PVS-API und auch web.impfnachweis.info wieder. Oder es ist eben instabil oder der Last nicht gewachsen. Das ist vielleicht ein Betriebsthema. Die API ist freigegeben und kann genutzt werden. Wir bitten die Anwender um Geduld und hoffen auf Entspannung in den nächsten Tagen. Immerhin funktioniert der Weg über die SMC-B gut. Vielen Dank.

**Reply by dirkexe** [reply]
Status: API weiterhin stabil verfügbar. Es funktioniert in den Arztpraxen.

**Reply by mge-medisoftware** [reply]
Ja, läuft meiste Zeit gut (je nach Auslastung Server). Praxen sind begeistert wie einfach das Ausstellen ist.

**Reply by ManuelB** [reply]
Wir konnten heute auch in der PU Zertifikate ausstellen.

Code: https://github.com/ere-health/dgc-ps-app

**Comment by dirkexe** [comment]
> ### PU Environment
> 
>     * 14.06.2021: Deployment of API Release 2.0.0
>     * 21.06.2021: Deployment of API Release 2.1.0

Wir erhalten Meldungen der Arztpraxen, die Genesenenzertifikate funktionieren noch nicht.
Wann können wir mit der für heute angekündigten Umstellung auf API Release 2.1.0 rechnen?

```
400 Bad Request
Sorry, there is something invalid in your request: Test and recoveries certificates are not allowed.
```

**Reply by thinkberg** [reply]
Guten Morgen. Es liegt noch keine Freigabe fuer dieses Feature vor. Sobald die Erfolgt, wird die Schnittstelle aktualisiert.

**Reply by fue-medatixx** [reply]
Hallo, noch ist ja der 21.06. - informieren Sie proaktiv wenn die Freigabe vorliegt und die Schnittstelle aktualisiert wurde? Das wäre hilfreich

**Reply by oliver-steinbrecher** [reply]
ja machen wir

**Comment by oliver-steinbrecher** [comment]
PU Update has been postponed 1 day and is going to happen tonight

**Comment by oliver-steinbrecher** [comment]
Production has been updated and support now test and recovery certificates

**Reply by dirkexe** [reply]
Test successfully: generation of recovery certificates works in PU. The API provides a PDF and/or CBOR.

**Reply by dirkexe** [reply]
I have tried to read a recovery certificate into the current iPhone apps:
CWA, CovPass Check and CovPass do not accept the QR code, but as I interpret it due to the lack of implementation of the standard.
Covid Cert (Switzerland), on the other hand, can import the recovery certificates, just does not accept the signature, but that's another issue.

In short: It works, but the apps now have to do their homework.

**Reply by oliver-steinbrecher** [reply]
Corresponding apps will be available in the next couple of Hours in the App Stores

**Reply by dirkexe** [reply]
With the just updated apps CovPass Check and CovPass, the QR codes from the PU are accepted.

**Reply by ruiyan-yu** [reply]
This bug is not fixed in PU: https://github.com/Digitaler-Impfnachweis/certification-apis/discussions/108#discussion-3414831

**Reply by ruiyan-yu** [reply]
now the vac-cert is broken too **IN THE PU**
![ERROR](https://user-images.githubusercontent.com/18198895/123044498-0b4fdc80-d3fa-11eb-90cf-e5dcefc7fd37.PNG)

**Reply by thinkberg** [reply]
fix is underway

**Comment by fue-medatixx** [comment]
> ### RU Environment
> * 31.05.2021: Deployment of API [Release 2.0.0](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/2.0.0)
> * 09.06.2021: Deployment of API [Release 2.1.0](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/2.1.0)
> 
> ### PU Environment
> (Regular maintenance windows: daily 20:00-21:00 UTC+2)
> 
> * 14.06.2021: Deployment of API [Release 2.0.0](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/2.0.0)
> * 22.06.2021: Deployment of API [Release 2.1.0](https://github.com/Digitaler-Impfnachweis/certification-apis/tree/2.1.0)

Hallo, die aktuelle Version in der OpenApi-Beschreibung ist die 2.2.0.
Hierbei ist wohl der **Header X-UBIRCH-DCCTYPE neu**.
**Wann geht diese Version denn in der RU bzw. PU live?**
Wenn man diesen Header jetzt schon mitschickt, wird er ignoriert oder führt das zum Fehler beim Request?

Für schnelle Antwort wären wir dankbar, wir stehen kurz vor unserem nächsten Kundenrelease und müssen wissen, was wir ausliefern können.

Besten Dank.

**Reply by thinkberg** [reply]
Wird im Moment noch nicht enforced und ist sinnvoll einzubinden. Aktuell wird der Header noch ignoriert fuer alle API Aufrufe, ausser die Hash API.

**Reply by fue-medatixx** [reply]
> Wird im Moment noch nicht enforced und ist sinnvoll einzubinden. Aktuell wird der Header noch ignoriert fuer alle API Aufrufe, ausser die Hash API.

Vielen Dank für die schnelle Antwort!
Gilt Gleiches auch für den Content-Type application/vnd.dgc.v1.3+json? Kann man diesen jetzt auch schon in PU verwenden?

**Reply by thinkberg** [reply]
Soweit ich weiss ja.
